### PR TITLE
[realm] non-actor handlers

### PR DESCRIPTION
**Pull request by SKNZ**
* close database connection in actor who opens it
* make login in account table unique
* add support for multiple realms (still in progress)
* add support for realms auto registration
* add shortcuts for databases (AuthDB, RealmDB)
* add RealmContext to provide information about current realm for
children of RealmServer
* rework handlers for AuthServer so that they're not actors
    * they are composed into AuthSession
* validate SHA hash in AuthSessionHandler
* rework handlers for RealmServer so that they're not actors
    * now they are objects that are being called in context by
    NetworkWorker/Session/SessionPlayer
    * HandledBy class says who should handle what
* rename some messages such that they tell more about their actions
    * also remove some "sealed trait Message" inheritance which was useless
* add CanSendPackets/ForwardToNetworkWorker traits to factorize packet
sending

### CODE REVIEW

File: wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala [hunk 1]
```diff
@@ -2,25 +2,70 @@ package wow.realm.session
...
-class NetworkWorker extends Actor with ActorLogging {
-  context.actorOf(Session.props, Session.PreferredName)
+class NetworkWorker(override implicit val realm: RealmContextData)
+  extends Actor
+          with ActorLogging
+          with PacketHandlerTag
+          with RealmContext
+          with CanSendPackets {
+
+  var _session: Option[ActorRef] = None
```
Comment (nyancol) [inline wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala]: Shouldn't it be private?

File: wow/core/src/main/scala/wow/realm/handlers/MovementHandler.scala [hunk 2]
```diff
@@ -2,60 +2,66 @@ package wow.realm.handlers
...
-  OpCodes.MoveFallReset,
-  OpCodes.MoveSetFly,
-  OpCodes.MsgMoveStartAscend,
-  OpCodes.MsgMoveStopAscend,
-  OpCodes.MoveChangeTransport,
-  OpCodes.MsgMoveStartDescend
-)
+object MovementHandlerCompanion {
+  val SupportedOpCodes: OpCodes.ValueSet = OpCodes.ValueSet(
+    OpCodes.MsgMoveStartForward,
```
Comment (nyancol) [inline wow/core/src/main/scala/wow/realm/handlers/MovementHandler.scala]: You could ignore "OpCodes.". Is there a reason why you kept it?

File: wow/core/src/main/scala/wow/realm/session/Session.scala [hunk 3]
```diff
@@ -1,27 +1,42 @@
...
-class Session extends Actor with ActorLogging {
-  private val networkWorker = context.parent
+class Session(val login: String, override val networkWorker: ActorRef)(override implicit val realm: RealmContextData)
+  extends Actor
+          with ActorLogging
+          with RealmContext
+          with PacketHandlerTag
+          with ForwardToNetworkWorker {
+
+  var player: Option[ActorRef] = _
```
Comment (nyancol) [inline wow/core/src/main/scala/wow/realm/session/Session.scala]: Is it useful?

File: wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala [hunk 4]
```diff
@@ -2,25 +2,70 @@ package wow.realm.session
...
+
+  def session_=(sessionRef: ActorRef): Unit = _session = {
+    if (_session.nonEmpty) {
+      throw new IllegalStateException("Session actor ref can only be set once")
+    }
+
+    Some(sessionRef)
+  }
+
+  var _player: Option[ActorRef] = None
```
Comment (nyancol) [inline wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala]: Shouldn't it be private?

File: wow/core/src/main/scala/wow/auth/AuthServer.scala [hunk 5]
```diff
@@ -32,21 +31,22 @@ class AuthServer extends Actor with ActorLogging {
...
-  private val eventRealmlist: EventRealmlist = EventRealmlist(serverRealmlistPacketBits)
+  private val realms = mutable.HashMap[Int, ServerRealmlistEntry]()
 
-  override def receive: PartialFunction[Any, Unit] = {
-    case GetRealmlist => sender() ! eventRealmlist
+  override def receive: Receive = {
+    case RegisterRealm(id, name, host, port) =>
+      log.debug(s"Realm added to list: $name at $host:$port")
+      realms(id) = ServerRealmlistEntry(1, 0, 0, name, s"$host:$port", 0, 1, 1, id)
+      AuthServer.realmlistPacketBits =
```
Comment (SKNZ) [inline wow/core/src/main/scala/wow/auth/AuthServer.scala]: The packet is the same for every player therefore re-serializing it every time is not necessary as we can just reuse the same bytes.
In the AuthServer, the packet is serialized every time a new realm is created. This should happen once or twice on startup at most.

For the future, this packet will be re-serialized every time the state of realm changes (i.e. if it crashes/is offline then the packet will be updated to reflect that) but we don't currently support that.

File: wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala (hunk 1 above)
Comment (SKNZ) [inline wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala]: Very good remark, fixed.

File: wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala (hunk 4 above)
Comment (SKNZ) [inline wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala]: Very good remark, fixed.

File: wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala [hunk 6]
```diff
@@ -115,59 +145,87 @@ class NetworkWorker extends Actor with ActorLogging {
...
 
-  case class EventTerminateWithPayload[A <: Payload with ServerSide](payload: A)
-    (implicit codec: Codec[A], opCodeProvider: OpCodeProvider[A]) extends PayloadBearingEvent[A](payload)
+  case class HandlePacket(header: ClientHeader, payloadBits: BitVector)
 
-  case class EventAuthenticated(sessionKey: BigInt) extends RealmSessionEvent
+}
+
+class NetworkWorkerFactory(implicit realm: RealmContextData) extends SessionActorCompanion {
+  override def props: Props = Props(new NetworkWorker()(realm))
```
Comment (SKNZ) [inline wow/core/src/main/scala/wow/realm/session/NetworkWorker.scala]: I think so. I've modified the code a little bit such that the syntax checker does not fail.

File: wow/core/src/main/scala/wow/realm/session/Session.scala (hunk 3 above)
Comment (SKNZ) [inline wow/core/src/main/scala/wow/realm/session/Session.scala]: Not as of now, but will definitely need it later.

File: wow/core/src/main/scala/wow/realm/handlers/MovementHandler.scala (hunk 2 above)
Comment (SKNZ) [inline wow/core/src/main/scala/wow/realm/handlers/MovementHandler.scala]: Not sure what you mean, I would need to import OpCodes._ for that to work no ? If that's what you're suggesting, I agree.

**Pull request merged**